modules/xmpp/JingleSessionPC.js

```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
+        remoteSdp.media[idx] += lines;
+    });
+    remoteSdp.raw = remoteSdp.session + remoteSdp.media.join('');
+
+    return remoteSdp;
+};
+
+/**
+ * Do a new o/a flow using the existing remote description
+ * @param sdp: optional remote sdp to use.  if not provided, the remote
+ *  sdp from the peer connection will be used
+ */
+JingleSessionPC.prototype._renegotiate = function(optionalRemoteSdp) {
+    let media_constraints = this.media_constraints;
+    let remoteSdp = optionalRemoteSdp || new SDP(this.peerconnection.remoteDescription.sdp);
+    let remoteDescription = new RTCSessionDescription({
+        type: "offer",
+        sdp: remoteSdp.raw
+    });
+
+    return new Promise((resolve, reject) => {
+        //TODO(brian): should this go elsewhere?
+        let remoteUfrag = getUfrag(remoteDescription.sdp);
+        if (remoteUfrag != this.remoteUfrag) {
+            this.remoteUfrag = remoteUfrag;
+            this.room.eventEmitter.emit(
+                    XMPPEvents.REMOTE_UFRAG_CHANGED, remoteUfrag);
+        }
+
+        this.peerconnection.setRemoteDescription(
+            remoteDescription,
+            () => {
+                if (this.signalingState === 'closed') {
+                    logger.error("Attempted to call create answer in closed");
+                    return;
+                }
+                this.peerconnection.createAnswer(
+                    (answer) => {
+                        //TODO(brian): should this go elsewhere?
+                        let localUfrag = getUfrag(answer.sdp);
+                        if (localUfrag != this.localUfrag) {
+                            this.localUfrag = localUfrag;
+                            this.room.eventEmitter.emit(
+                                    XMPPEvents.LOCAL_UFRAG_CHANGED, localUfrag);
+                        }
+                        this.peerconnection.setLocalDescription(
+                            answer,
+                            () => { resolve(); },
+                            (error) => { reject(error, "setLocalDescription failed"); }
+                        );
+                    },
+                    (error) => { reject(error, "createAnswer failed"); },
+                    media_constraints
+                );
+            },
+            (error) => { reject(error, "setRemoteDescription failed"); }
+        );
+    });
+};
+
+/**
+ * Replaces oldStream with newStream and performs a single offer/answer
+ *  cycle after both operations are done.  Either oldStream or newStream
+ *  can be null; replacing a valid 'oldStream' with a null 'newStream'
+ *  effectively just removes 'oldStream'
+ * @param oldStream the current stream in use to be replaced
+ * @param newStream the new stream to use
+ * @returns {Promise}
+ */
+JingleSessionPC.prototype.replaceStream = function (oldStream, newStream) {
+    return new Promise((resolve, reject) => {
+        let workFunction = (finishedCallback) => {
```

Comment:
same as my comment above...i just think it's easier to read and is more self-documenting with the 'workFunction' name.  i can change it though.